cpp/test/core/maybe_tests.cpp

```diff
@@ -107,6 +126,259 @@ TEST_CASE_BEGIN(MaybeInterface)
 }
 TEST_CASE_END
 
+TEST_CASE_BEGIN(CopyAndMoveFromValues)
+{
+    { // simple type
+        int i = 100;
+        bond::maybe<int> copiedValue{ i };
+        BOOST_REQUIRE(copiedValue);
+        BOOST_CHECK(copiedValue.value() == 100);
+
+        copiedValue = 200;
+        BOOST_REQUIRE(copiedValue);
+        BOOST_CHECK(copiedValue.value() == 200);
+    }
+
+    { // move-only type
+        std::unique_ptr<int> up{ new int{ 100 } };
+        bond::maybe<std::unique_ptr<int>> movedFromValue{ std::move(up) };
+        BOOST_CHECK(!up);
+        BOOST_REQUIRE(movedFromValue);
+        BOOST_CHECK(*movedFromValue.value() == 100);
+
+        up.reset(new int{ 200 });
+        movedFromValue = std::move(up);
+        BOOST_CHECK(!up);
+        BOOST_REQUIRE(movedFromValue);
+        BOOST_CHECK(*movedFromValue.value() == 200);
+    }
+
+    { // type with allocator
+        auto numbers = { 1, 2, 3, 4, 5 };
+        auto primes = { 2, 3, 5 };
+        std::vector<int> vecNumbers{ std::begin(numbers), std::end(numbers) };
+        std::vector<int> vecPrimes{ std::begin(primes), std::end(primes) };
+
+        bond::maybe<std::vector<int>> copiedValue{ vecNumbers };
+        BOOST_REQUIRE(copiedValue);
+        BOOST_CHECK(copiedValue.value() == vecNumbers);
+
+        copiedValue = vecPrimes;
+        BOOST_REQUIRE(copiedValue);
+        BOOST_CHECK(copiedValue.value() == vecPrimes);
+
+        bond::maybe<std::vector<int>> movedValue{ std::move(vecNumbers) };
+        BOOST_CHECK(vecNumbers.empty());
+        BOOST_REQUIRE(movedValue);
+        BOOST_CHECK(movedValue.value().size() == 5);
+
+        movedValue = std::move(vecPrimes);
+        BOOST_CHECK(vecPrimes.empty());
+        BOOST_REQUIRE(movedValue);
+        BOOST_CHECK(movedValue.value().size() == 3);
+    }
+
+    { // move-only type with allocator
+        std::vector<std::unique_ptr<int>> source1;
+        source1.emplace_back(new int{ 1 });
+
+        bond::maybe<std::vector<std::unique_ptr<int>>> movedValue{ std::move(source1) };
+        BOOST_CHECK(source1.empty());
+        BOOST_REQUIRE(movedValue);
+        BOOST_CHECK(movedValue.value().size() == 1);
+
+        std::vector<std::unique_ptr<int>> source2;
+        source2.emplace_back(new int{ 2 });
+        source2.emplace_back(new int{ 2 });
+
+        movedValue = std::move(source2);
+        BOOST_CHECK(source2.empty());
+        BOOST_REQUIRE(movedValue);
+        BOOST_CHECK(movedValue.value().size() == 2);
+    }
+}
+TEST_CASE_END
+
```

Comment:
Test for non-default-constructible types?

Reply:
The `AllocatorPropagated` test below uses a non-default-constructible type (with an allocator). Are you saying we need a test for a non-default-constructible type without an allocator?

Reply:
No, that should be enough.